File: tests/ClaudeToolbar.Core.Tests/Mascot/MascotModelBuilderTests.cs
using ClaudeToolbar.Core.Formatting;
using ClaudeToolbar.Core.Mascot;
using ClaudeToolbar.Core.Widget;

namespace ClaudeToolbar.Core.Tests.Mascot;

public class MascotModelBuilderTests
{
    private static WidgetModel Widget(bool dimmed = false, params BarLevel[] levels) =>
        new(levels.Select((l, i) => new WidgetRow($"r{i}", 10, "10%", "", l)).ToList(), dimmed, false, null);

    [Fact]
    public void OffModeHidesTheMascot() =>
        Assert.Equal(MascotModel.Hidden, MascotModelBuilder.Build(Widget(false, BarLevel.Ok), "off", 12));

    [Fact]
    public void NoticeWithoutRowsHidesTheMascot() =>
        Assert.Equal(MascotModel.Hidden, MascotModelBuilder.Build(new WidgetModel([], false, false, "Sign in with claude"), "full", 12));

    [Fact]
    public void WorstLevelWinsAndArmAnglePassesThrough()
    {
        var m = MascotModelBuilder.Build(Widget(false, BarLevel.Ok, BarLevel.Crit, BarLevel.Warn), "full", 33);
        Assert.True(m.Visible);
        Assert.Equal(BarLevel.Crit, m.Level);
        Assert.Equal(33, m.ArmAngle);
        Assert.False(m.Dimmed);
    }

    [Fact]
    public void HoverModeIsVisibleAndDimmedFollowsTheWidget()
    {
        var m = MascotModelBuilder.Build(Widget(true, BarLevel.Ok), "hover", WaveAnimation.RestAngle);
        Assert.True(m.Visible);
        Assert.Equal(BarLevel.Ok, m.Level);
        Assert.True(m.Dimmed);
    }

    [Fact]
    public void UnknownModeBehavesLikeFull() =>
        Assert.True(MascotModelBuilder.Build(Widget(false, BarLevel.Ok), "sparkles", 0).Visible);
}

File: mac/Tests/ClaudeToolbarCoreTests/Mascot/MascotModelBuilderTests.swift
import XCTest
@testable import ClaudeToolbarCore

final class MascotModelBuilderTests: XCTestCase {
    private func status(dimmed: Bool = false, _ levels: BarLevel...) -> StatusItemModel {
        StatusItemModel(rows: levels.enumerated().map { i, level in
            StatusRow(label: "r\(i)", utilization: 10, percentText: "10%", timeText: "", level: level, hasData: true)
        }, dimmed: dimmed, showStaleDot: false, notice: nil, hint: nil)
    }

    func testOffModeHidesTheMascot() {
        XCTAssertEqual(MascotModelBuilder.build(status: status(.ok), mascotMode: "off", armAngle: 12), .hidden)
    }

    func testNoticeWithoutRowsHidesTheMascot() {
        let notice = StatusItemModel(rows: [], dimmed: false, showStaleDot: false, notice: "Sign in with claude", hint: nil)
        XCTAssertEqual(MascotModelBuilder.build(status: notice, mascotMode: "full", armAngle: 12), .hidden)
    }

    func testWorstLevelWinsAndArmAnglePassesThrough() {
        let m = MascotModelBuilder.build(status: status(.ok, .crit, .warn), mascotMode: "full", armAngle: 33)
        XCTAssertTrue(m.visible)
        XCTAssertEqual(m.level, .crit)
        XCTAssertEqual(m.armAngle, 33)
        XCTAssertFalse(m.dimmed)
    }

    func testHoverModeIsVisibleAndDimmedFollowsTheWidget() {
        let m = MascotModelBuilder.build(status: status(dimmed: true, .ok), mascotMode: "hover", armAngle: WaveAnimation.restAngle)
        XCTAssertTrue(m.visible)
        XCTAssertEqual(m.level, .ok)
        XCTAssertTrue(m.dimmed)
    }

    func testUnknownModeBehavesLikeFull() {
        XCTAssertTrue(MascotModelBuilder.build(status: status(.ok), mascotMode: "sparkles", armAngle: 0).visible)
    }
}
